File: C++/Bridge/ColorfulShapes/shape.cc
#include <iostream>
#include "shape.h"

Shape::Shape(std::unique_ptr<Color> color) : color{std::move(color)} {
}

void Shape::set_color(std::unique_ptr<Color> color) {
    this->color = std::move(color);
}


Rectangle::Rectangle(std::unique_ptr<Color> color) : Shape(std::move(color)) {
}

void Rectangle::draw_shape() const {
    std::cout << "Drawing a rectangle...\n";
    color->fill_color();
}

void Rectangle::resize(double percent) const {
    std::cout << "Resizing the rectangle to " << percent << " % of its original size...\n";
}


Triangle::Triangle(std::unique_ptr<Color> color) : Shape(std::move(color)) {
}

void Triangle::draw_shape() const {
    std::cout << "Drawing a triangle...\n";
    color->fill_color();
}

void Triangle::resize(double percent) const {
    std::cout << "Resizing the triangle to " << percent << " % of its original size...\n";
}

File: Java/Bridge/ColorfulShapes/src/com/ericgtkb/Shape.java
package com.ericgtkb;

// Without the bridge pattern, we will need (number of shapes * number of colors) classes and
// the Color classes and Shape classes cannot vary independently
public abstract class Shape {
    protected Color color;

    public Shape(Color color) {
        this.color = color;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    public abstract void drawShape();
    public abstract void resize(double percent);

}

class Rectangle extends Shape {
    public Rectangle(Color color) {
        super(color);
    }

    @Override
    public void drawShape() {
        System.out.println("Drawing a rectangle...");
        color.fillColor();
    }

    @Override
    public void resize(double percent) {
        System.out.println("Resizing the rectangle to " + percent + " % of its original size");
    }
}


class Triangle extends Shape {
    public Triangle(Color color) {
        super(color);
    }

    @Override
    public void drawShape() {
        System.out.println("Drawing a triangle...");
        color.fillColor();
    }

    @Override
    public void resize(double percent) {
        System.out.println("Resizing the triangle to " + percent + " % of its original size");
    }
}
